alg/viewshed/cumulative_viewshed.cpp

```diff
@@ -0,0 +1,196 @@
+/// @return  True on success, false otherwise.
+bool Cumulative::run(const std::string &srcFilename,
+                     GDALProgressFunc pfnProgress, void *pProgressArg)
+{
+    // In cumulative view, we run the executors in normal mode and want "1" where things
+    // are visible.
+    m_opts.outputMode = OutputMode::Normal;
+    m_opts.visibleVal = 1;
+
+    //ABELL
+    /**
+    if (!setupProgress(pfnProgress, pProgressArg))
+        return false;
+    **/
+    DatasetPtr srcDS(
+        GDALDataset::FromHandle(GDALOpen(srcFilename.c_str(), GA_ReadOnly)));
+    GDALRasterBand *pSrcBand = srcDS->GetRasterBand(1);
+
+    // In cumulative mode, the output extent is always the entire source raster.
+    m_extent.xStop = GDALGetRasterBandXSize(pSrcBand);
+    m_extent.yStop = GDALGetRasterBandYSize(pSrcBand);
+
+    // Make a bunch of observer locations based on the spacing and stick them on a queue
+    // to be handled by viewshed executors.
+    for (int x = 0; x < m_extent.xStop; x += m_opts.observerSpacing)
+        for (int y = 0; y < m_extent.yStop; y += m_opts.observerSpacing)
+            m_observerQueue.push({x, y});
+    m_observerQueue.done();
+
+    // Run executors.
+    const int numThreads = NUM_JOBS;
+    std::atomic<bool> err = false;
+    std::atomic<int> running = numThreads;
+    CPLWorkerThreadPool executorPool(numThreads);
+    for (int i = 0; i < numThreads; ++i)
+        executorPool.SubmitJob([this, &srcFilename, &err, &running]
+                               { runExecutor(srcFilename, err, running); });
+
+    // Run combiners that create 8-bit sums of executor jobs.
+    CPLWorkerThreadPool combinerPool(numThreads);
+    std::vector<Combiner> combiners(numThreads,
+                                    Combiner(m_datasetQueue, m_bufQueue));
+    for (Combiner &c : combiners)
+        combinerPool.SubmitJob([&c] { c.run(); });
+
+    // Run 32-bit rollup job.
+    std::thread sum([this] { rollupRasters(); });
+
+    // When the combiner jobs are done, all the data is in the buf queue.
+    combinerPool.WaitCompletion();
+    if (m_datasetQueue.isStopped())
+        return false;
+    m_bufQueue.done();
+
+    // Wait for finalBuf to be fully filled. Then scale the output data.
+    sum.join();
+    scaleOutput();
+    bool ok = writeOutput(*pSrcBand);
+
+    executorPool.WaitCompletion();
+    (void)pfnProgress;
+    (void)pProgressArg;
+    return ok;
+}
+
+void Cumulative::runExecutor(const std::string &srcFilename,
+                             std::atomic<bool> &err, std::atomic<int> &running)
+{
+    Location loc;
+    while (!err && m_observerQueue.pop(loc))
+    {
+        DatasetPtr srcDs(GDALDataset::Open(srcFilename.c_str(), GA_ReadOnly));
+        if (!srcDs)
+        {
+            err = true;
+            break;
+        }
+
+        GDALDriver *memDriver = GetGDALDriverManager()->GetDriverByName("MEM");
+        DatasetPtr dstDs(memDriver->Create(
```

Comment:
dstDs should be checked against nullptr (memory allocation failure)

Reply:
So cumulative viewshed requires a temporary in-memory raster of the size of the source raster ? That should perhaps be indicated in the documentation as a limitation.